File: listeners/commands/pr_settings.py
from slack_sdk.errors import SlackApiError
from database_settings import get_channel_sla_time, get_channel_enabled_hours, set_channel_sla_time, toggle_channel_hour
from helpers import is_valid_int

def pr_settings_callback(ack, body, client, logger):
    ack()
    channel_id = body['channel_id']
    
    # Retrieve the current settings for this channel (SLA time and enabled hours)
    current_sla_time = get_channel_sla_time(channel_id)  # Default to 8 if not set
    enabled_hours = get_channel_enabled_hours(channel_id)  # Default to all enabled
    logger.info(f"PR settings callback invoked for channel {channel_id}. Current SLA time: {current_sla_time}, Enabled hours: {enabled_hours}")

    # Build the modal view
    modal_view = {
        "type": "modal",
        "callback_id": "pr_settings_modal",
        "private_metadata": channel_id,  # Pass channel_id to use later
        "title": {
            "type": "plain_text",
            "text": "PR Settings"
        },
        "submit": {
            "type": "plain_text",
            "text": "Save"
        },
        "close": {
            "type": "plain_text",
            "text": "Cancel"
        },
        "blocks": [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": "*Configure settings for this channel:*"
                }
            },
            {
                "type": "input",
                "block_id": "sla_time_input",
                "element": {
                    "type": "plain_text_input",
                    "action_id": "sla_time_input_action",
                    "initial_value": str(current_sla_time),
                    "placeholder": {
                        "type": "plain_text",
                        "text": "Enter SLA Time in hours"
                    }
                },
                "label": {
                    "type": "plain_text",
                    "text": "SLA Time (in hours)"
                }
            },
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": "Configure SLA Check Hours PST (click to enable/disable):"
                }
            }
        ]
    }

    # Add buttons for each hour (9 AM to 4 PM PST)
    action_blocks = []
    for hour in range(9, 17):  # 9 AM to 4 PM
        format_12_hours = hour
        if format_12_hours > 12:
            format_12_hours -= 12
        button_block = {
            "type": "button",
            "text": {
                "type": "plain_text",
                "text": f"{format_12_hours}:00"
            },
            "action_id": f"toggle_hour_{hour}"
        }
        # Only set "primary" style if the hour is enabled
        if hour in enabled_hours:
            button_block["style"] = "primary"
        action_blocks.append(button_block)

    # Add action buttons in a row
    modal_view["blocks"].append({
        "type": "actions",
        "block_id": "hour_buttons",
        "elements": action_blocks
    })

    # Open the modal
    try:
        client.views_open(
            trigger_id=body['trigger_id'],
            view=modal_view
        )
        logger.info(f"Modal for PR settings opened successfully for channel {channel_id}")
    except SlackApiError as e:
        logger.error(f"Error opening modal for channel {channel_id}: {e.response['error']}")

def handle_sla_time_input(ack, body, client, logger):
    ack()
    
    # Extract channel_id from private_metadata
    channel_id = body['view']['private_metadata']
    
    # Extract SLA time input from the modal submission
    sla_time = body['view']['state']['values']['sla_time_input']['sla_time_input_action']['value']

    # Check if the SLA time is a valid integer
    if not is_valid_int(sla_time):
        ack(response_action="errors", errors={
            "sla_time_input": "Please enter an integer"
        })
        logger.warning(f"Invalid SLA time input for channel {channel_id}: {sla_time}")
        return
    
    try:
        # Store the new SLA time for this channel
        set_channel_sla_time(channel_id, sla_time)
        logger.info(f"SLA time for channel {channel_id} updated to {sla_time} hours.")
    except Exception as e:
        logger.error(f"Error updating SLA time for channel {channel_id}: {e}")

def handle_toggle_hour(ack, body, client, logger):
    # Acknowledge the button click and update the blocks dynamically
    ack(response_action="update", view={
        "type": body['view']['type'],
        "callback_id": body['view']['callback_id'],
        "title": body['view']['title'],
        "blocks": body['view']['blocks'],
        "private_metadata": body['view']['private_metadata'],
    })

    action_id = body['actions'][0]['action_id']
    hour = int(action_id.split("_")[-1])
    channel_id = body['view']['private_metadata']
    
    try:
        # Toggle the enabled state of the hour for the channel
        toggle_channel_hour(channel_id, hour)
        enabled_hours = get_channel_enabled_hours(channel_id)
        logger.info(f"Hour {hour}:00 toggled for channel {channel_id}. Enabled hours: {enabled_hours}")

        # Update the button style based on the new state
        for block in body['view']['blocks']:
            if block['block_id'] == "hour_buttons":
                for element in block['elements']:
                    if element['action_id'] == action_id:
                        if hour in enabled_hours:
                            element['style'] = "primary"  # Set the primary style if the hour is enabled
                        else:
                            # Remove the style attribute if the hour is disabled
                            if 'style' in element:
                                del element['style']
        
        # Respond with updated button states without re-opening the modal
        client.views_update(
            view_id=body['view']['id'],
            view={
                "type": body['view']['type'],
                "callback_id": body['view']['callback_id'],
                "title": body['view']['title'],
                "blocks": body['view']['blocks'],
                "private_metadata": body['view']['private_metadata'],
                "submit": {
                    "type": "plain_text",
                    "text": "Save"
                },
                "close": {
                    "type": "plain_text",
                    "text": "Cancel"
                }
            }
        )
        logger.info(f"View updated successfully for channel {channel_id}, hour {hour}:00.")
    except Exception as e:
        logger.error(f"Error toggling hour {hour} for channel {channel_id}: {e}")

File: src/listeners/commands/PrSettings.java
package prreminderbot.actions;

import com.slack.api.bolt.context.builtin.ActionContext;
import com.slack.api.methods.SlackApiException;
import com.slack.api.Slack;
import com.slack.api.model.block.Blocks;
import com.slack.api.model.block.element.BlockElement;
import com.slack.api.model.block.element.BlockElements;
import com.slack.api.model.view.Views;
import org.slf4j.Logger;
import prreminderbot.database.DatabaseChannelSettings;
import prreminderbot.helpers.Helpers;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class PrSettingsCallback {

    public static void handlePrSettingsCallback(ActionRequest req, ActionContext ctx, Slack client, Logger logger) {
        try {
            ctx.ack();

            String channelId = req.getPayload().getChannel().getId();

            // Retrieve the current settings for this channel (SLA time and enabled hours)
            int currentSlaTime = DatabaseChannelSettings.getChannelSlaTime(channelId);  // Default to 8 if not set
            List<Integer> enabledHours = DatabaseChannelSettings.getChannelEnabledHours(channelId);  // Default to all enabled
            logger.info("PR settings callback invoked for channel {}. Current SLA time: {}, Enabled hours: {}", channelId, currentSlaTime, enabledHours);

            // Build the modal view
            List<BlockElement> actionBlocks = new ArrayList<>();
            for (int hour = 9; hour <= 16; hour++) {  // 9 AM to 4 PM
                int displayHour = hour > 12 ? hour - 12 : hour;
                BlockElement buttonBlock = BlockElements.button(b -> b
                    .text(pt -> pt.type("plain_text").text(displayHour + ":00"))
                    .actionId("toggle_hour_" + hour)
                );
                if (enabledHours.contains(hour)) {
                    buttonBlock = BlockElements.button(b -> b
                        .text(pt -> pt.type("plain_text").text(displayHour + ":00"))
                        .actionId("toggle_hour_" + hour)
                        .style("primary")  // Set "primary" style if the hour is enabled
                    );
                }
                actionBlocks.add(buttonBlock);
            }

            // Build the modal view blocks
            List<com.slack.api.model.block.Block> blocks = new ArrayList<>();
            blocks.add(Blocks.section(section -> section.text(text -> text.type("mrkdwn").text("*Configure settings for this channel:*"))));
            blocks.add(Blocks.input(input -> input
                .blockId("sla_time_input")
                .element(BlockElements.plainTextInput(ti -> ti.actionId("sla_time_input_action").initialValue(String.valueOf(currentSlaTime))))
                .label(label -> label.type("plain_text").text("SLA Time (in hours)"))
            ));
            blocks.add(Blocks.section(section -> section.text(text -> text.type("mrkdwn").text("*Configure SLA Check Hours PST (click to enable/disable)*"))));
            blocks.add(Blocks.actions(actions -> actions.blockId("hour_buttons").elements(actionBlocks)));

            // Open the modal
            client.methods().viewsOpen(r -> r
                .triggerId(req.getPayload().getTriggerId())
                .view(Views.view(view -> view
                    .type("modal")
                    .callbackId("pr_settings_modal")
                    .privateMetadata(channelId)
                    .title(vt -> vt.type("plain_text").text("PR Settings"))
                    .blocks(blocks)
                    .submit(s -> s.type("plain_text").text("Save"))
                    .close(c -> c.type("plain_text").text("Cancel"))
                ))
            );
            logger.info("Modal for PR settings opened successfully for channel {}", channelId);

        } catch (IOException | SlackApiException e) {
            logger.error("Error opening modal for channel {}: {}", req.getPayload().getChannel().getId(), e.getMessage());
        }
    }

    public static void handleSlaTimeInput(ActionRequest req, ActionContext ctx, Slack client, Logger logger) {
        try {
            ctx.ack();

            String channelId = req.getPayload().getView().getPrivateMetadata();
            String slaTime = req.getPayload().getView().getState().getValues().get("sla_time_input").get("sla_time_input_action").getValue();

            // Validate SLA time input
            if (!Helpers.isValidInt(slaTime)) {
                ctx.ack(r -> r.responseAction("errors").errors("sla_time_input", "Please enter a valid integer"));
                logger.warn("Invalid SLA time input for channel {}: {}", channelId, slaTime);
                return;
            }

            // Update the SLA time for the channel
            DatabaseChannelSettings.setChannelSlaTime(channelId, Integer.parseInt(slaTime));
            logger.info("SLA time for channel {} updated to {} hours.", channelId, slaTime);

        } catch (IOException | SlackApiException e) {
            logger.error("Error updating SLA time for channel {}: {}", req.getPayload().getView().getPrivateMetadata(), e.getMessage());
        }
    }

    public static void handleToggleHour(ActionRequest req, ActionContext ctx, Slack client, Logger logger) {
        try {
            ctx.ack(r -> r.responseAction("update"));

            String channelId = req.getPayload().getView().getPrivateMetadata();
            int hour = Integer.parseInt(req.getPayload().getActions().get(0).getActionId().split("_")[2]);

            // Toggle the enabled state of the hour for the channel
            DatabaseChannelSettings.toggleChannelHour(channelId, hour);
            List<Integer> enabledHours = DatabaseChannelSettings.getChannelEnabledHours(channelId);
            logger.info("Hour {}:00 toggled for channel {}. Enabled hours: {}", hour, channelId, enabledHours);

            // Update the view with the new button styles
            List<com.slack.api.model.block.Block> updatedBlocks = req.getPayload().getView().getBlocks();
            for (com.slack.api.model.block.Block block : updatedBlocks) {
                if (block instanceof com.slack.api.model.block.ActionsBlock) {
                    for (BlockElement element : ((com.slack.api.model.block.ActionsBlock) block).getElements()) {
                        if (((com.slack.api.model.block.element.ButtonElement) element).getActionId().equals("toggle_hour_" + hour)) {
                            if (enabledHours.contains(hour)) {
                                ((com.slack.api.model.block.element.ButtonElement) element).setStyle("primary");
                            } else {
                                ((com.slack.api.model.block.element.ButtonElement) element).setStyle(null);  // Remove the style if the hour is disabled
                            }
                        }
                    }
                }
            }

            // Update the view
            client.methods().viewsUpdate(r -> r
                .viewId(req.getPayload().getView().getId())
                .view(Views.view(view -> view
                    .type(req.getPayload().getView().getType())
                    .callbackId(req.getPayload().getView().getCallbackId())
                    .title(req.getPayload().getView().getTitle())
                    .blocks(updatedBlocks)
                    .privateMetadata(channelId)
                    .submit(s -> s.type("plain_text").text("Save"))
                    .close(c -> c.type("plain_text").text("Cancel"))
                ))
            );
            logger.info("View updated successfully for channel {}, hour {}:00.", channelId, hour);

        } catch (IOException | SlackApiException e) {
            logger.error("Error toggling hour {} for channel {}: {}", req.getPayload().getActions().get(0).getActionId().split("_")[2], req.getPayload().getView().getPrivateMetadata(), e.getMessage());
        }
    }
}
